python/triton/experimental/gluon/language/_core.py

```diff
@@ -313,3 +314,8 @@ def warp_specialize(args, default_partition, worker_partitions, worker_num_warps
     worker_num_regs = [_unwrap_if_constexpr(r) for r in worker_num_regs]
     return _semantic.warp_specialize(args, default_partition, worker_partitions, worker_num_warps,  #
                                      worker_num_regs, _generator)
+
+
+@builtin
+def thread_barrier(_semantic=None):
```

Comment:
passby comment: we will have to define the right semantic for this at some point as the threads don't really exist at the Gluon level. 
I guess they do exist in layouts but barriers are meant to be inserted by the compiler, this is not true in few cases like global load/stores though

Reply:
but in triton, the semantic is closer to a fence. 
We don't want to get to a point where user have to insert all the barriers on their own so that means we need to define clear rules for when barriers needs to be inserted by users. We still see it as block programming where user doesn't have to worry about how to convert to SIMT

Reply:
Fair point, should we add an explicit fence operation to triton?